internal/nginx/config/servers.go

```diff
@@ -205,6 +205,11 @@ func createReturnValForRedirectFilter(filter *v1beta1.HTTPRequestRedirectFilter,
 	scheme := "$scheme"
 	if filter.Scheme != nil {
 		scheme = *filter.Scheme
+		// Don't specify the port in the return url if the scheme is
+		// well known and the port is already set to the correct well known port
+		if (port == 80 && scheme == "http") || (port == 443 && scheme == "https") {
+			hostnamePort = hostname
+		}
 		if filter.Port == nil {
```

Comment:
I don't think this case (filter.Port == nil ) is necessary, because by this line, port will be either filter.Port or listenerPort. An alternative suggestion is here -- https://github.com/nginxinc/nginx-kubernetes-gateway/pull/801#discussion_r1247019363

Reply:
Yes but consider the following case:

filter.Port is unset
filter.Schema is set to http
listenerPort is 8080

`port == 80 && scheme == "http"` will not evaluate to true, so the port 8080 will be added to the return, which contradicts the following:

> 	// If no port is specified, the redirect port MUST be derived using the
	// following rules:
	//
	// * If redirect scheme is not-empty, the redirect port MUST be the well-known
	//   port associated with the redirect scheme. Specifically "http" to port 80
	//   and "https" to port 443. If the redirect scheme does not have a
	//   well-known port, the listener port of the Gateway SHOULD be used.
	
In other words no port, scheme = http, listenerPort = 8080, url should not include the port. I either need to define the ports when scheme is set above to keep the hierarchy, or ignore the port if the scheme is specified but port is not specified. I thought checking for nil port again was cleaner, than defining the port to that of the scheme to just ignore it.

Reply:
> Yes but consider the following case:

got it. thanks for clarifying!